### PR TITLE
WIP: Broker Relist Subresource

### PR DESCRIPTION
**Pull request by eriknelson**
2nd PR implementing a relist subresource, as outlined by #1086.

### CODE REVIEW

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 1]
```diff
@@ -86,6 +87,14 @@ const (
...
 	// relisted when the spec of the broker changes.
 	ServiceBrokerRelistBehaviorManual ServiceBrokerRelistBehavior = "Manual"
+
+	// ServiceBrokerRelistRequestsMax is the maximum allowable value for
+	// RelistRequests before it is wrapped around.
+	ServiceBrokerRelistRequestsMax int64 = util.MaxInt64
+
+	// ServiceBrokerRelistRequestsMin is the minimum value that RelistRequests
+	// will be wrapped around to after reaching its maximum
+	ServiceBrokerRelistRequestsMin int64 = util.MinUint64
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: You need to add a note to the RelistRequests field explaining the wrapping behavior.

File: pkg/registry/servicecatalog/broker/strategy.go [hunk 2]
```diff
@@ -170,3 +179,45 @@ func (brokerStatusRESTStrategy) ValidateUpdate(ctx genericapirequest.Context, ne
...
+	newServiceBroker.ObjectMeta = oldServiceBroker.ObjectMeta
+
+	var nextRelistRequest int64
+	if oldServiceBroker.Spec.RelistRequests == sc.ServiceBrokerRelistRequestsMax {
+		nextRelistRequest = sc.ServiceBrokerRelistRequestsMin
+	} else {
+		nextRelistRequest = oldServiceBroker.Spec.RelistRequests + 1
+	}
+
+	newServiceBroker.Spec.RelistRequests = nextRelistRequest
```
Comment (pmorie) [inline pkg/registry/servicecatalog/broker/strategy.go]: Does this bump the generation?

File: test/integration/clientset_test.go [hunk 3]
```diff
@@ -353,6 +353,15 @@ func testBrokerClient(sType server.StorageType, client servicecatalogclient.Inte
...
 	}
 
+	// Assert incremented relistRequest on Relist subresource request
+	oldRelistRequests = brokerServer.Spec.RelistRequests
+	expectedRelistRequests = oldRelistRequests + 1
+	returnedBroker, err := brokerClient.Relist(brokerServer)
+	if e, a := expectedRelistRequests, returnedBroker.Spec.RelistRequests; !reflect.DeepEqual(e, a) {
+		return fmt.Errorf("Didn't get incremented relistRequests follow Relist: \nexpected: %v\n\ngot: %v", e, a)
+	}
+	// TODO ERIK: Test wrap? Need to set initial to max, confirm back to min...
```
Comment (pmorie) [inline test/integration/clientset_test.go]: Is this still a TODO?

**Comment by enj** [comment]
At some point I would like to discuss this with you guys, mostly because I feel like you are doing it wrong.

**Comment by duglin** [comment]
@enj can you elaborate?

**Comment by enj** [comment]
Sorry did not have time earlier to elaborate.  So, to be clear, I am not commenting on the validity of actually having a `relist` subresource.  I will admit that it strange to have an endpoint the exists solely to cause a change to an API type, and I do not really understand why you need or want to do that.  But I will leave that for you guys to decide as I lack the context to say if its a good idea or not.

---

So what I believe from this is a bad idea is the actual implementation of the `relist` subresource endpoint.  This implementation re-uses the registry store struct, which is fine if you need all the power and flexibility that gives you.  But you do not need that at all in this case (and you are offloading too much onto the user wanting to relist by forcing them to give you the broker resource as input).  And using this struct is dangerous because it tightly couples you to the underlying implementation, is effectively direct to etcd, and allows you to bypass parts of the standard REST path.  We made the mistake of doing this in a lot of the API types in Openshift.  And we realized it was a really bad idea, and have now gone through the pain to rewriting those bits to use the REST client interfaces.  So my suggestion would be for you to build this using the broker REST client interface.

Here is a rough draft of what I think you should do instead:

```go
type RelistBroker struct {
	metav1.TypeMeta
	metav1.ObjectMeta

	// Spec holds information about the request being evaluated.
	Spec RelistBrokerSpec

	// Status is filled in by the server and indicates whether the request succeeded
	Status RelistBrokerStatus
}

type RelistBrokerSpec struct {
	// you may not need this at all, but I could see this being something
	// like "relist" if you want to support other actions in the future
	// you would want more generic names for the struct if you want to
	// support such a use case though
	Action string
}

type RelistBrokerStatus struct {
	// did the relist happen?
	Success bool

	// if it failed, why did it fail
	Reason string
}

type relistBrokerREST struct {
	client brokerRESTInterfaceClient
}

func (r *relistBrokerREST) New() runtime.Object {
	return &RelistBroker{}
}

func (r *relistBrokerREST) Create(ctx genericapirequest.Context, name string, obj runtime.Object, includeUninitialized bool) (runtime.Object, error)
	rb, ok := obj.(*RelistBroker)
	if !ok {
		return nil, apierrors.NewBadRequest(fmt.Sprintf("not a RelistBroker: %#v", obj))
	}
	// perform any validation on rb as needed
	...
	// get the broker we are a sub resource of so we can relist it
	r.client.Get(name)
	// increment the counter with whatever logic you want
	...
	// update the broker
	err := r.client.Update(name)
	// retry on conflict if needed
	// create a new RB as the response (and fill in data you know to be true)
	rbOut := &RelistBroker{}
	rbOut.Status.Success = true
	return rbOut, nil
}
```

I find this approach to be significantly simpler.  It is basically what a user might do by hand, but turned into a very tightly controlled API.  Note that it requires no strategy because it uses a REST client, and it is not tightly coupled with the storage implementation of the broker itself.  If desired, instead of doing a `relist` subresource, you could simply do an `action` subresource that could later support different controlled actions against the broker.  But maybe you do not need that complexity.

**Comment by pmorie** [comment]
Sublist resource may have to be 0.1.1 item.

**Comment by pmorie** [comment]
@enj 

> I will admit that it strange to have an endpoint the exists solely to cause a change to an API type, and I do not really understand why you need or want to do that.

The purpose of this endpoint is to provide a REST api that results in the generation of the broker resource being bumped without requiring the user having to know how to change the spec in just the right way to cause a relist, or worse, making a change to the spec in another area just to cause a relist.

I will be honest here: we already have the functionality required for users to get the behavior we want, and the relist subresource isn't strictly required to do it.  So, if you would like to sherpa your preferred design through, that's fine.  If we can't agree on an approach, that's also fine.  In either event, I don't consider this to be in the critical path for a beta release, and have other things that are higher priority to give attention to.

So, in conclusion: @enj and @eriknelson, feel free to work together on this.  I suggest that you write up a proposal that shows examples of the body that should be sent to this call and the response returned to the user before you work further on implementation, and present that as a separate PR or issue.

**Comment by eriknelson** [comment]
I have some unrelated work I need to take care of this week, but I'm planning on submitting a proposal for an alternate implementation that incorporates @enj's suggestions to #1086 soon.

**Comment by pmorie** [comment]
@eriknelson starting this back up since it's been stalled for a while - I think @enj's comment here is ringing the bell for me.

@enj I also think that your example is a very concise one that would be useful to write up in an API doctrine-type document somewhere.  (related: kubernetes/community#892)

**Comment by eriknelson** [comment]
This is so far out of date I'm going to close.
